### ISSUE TITLE
`before` hook runs before every test in `describe` block instead of once 

### ISSUE THREAD

**Issue by elenasch** [cypress-io/cypress#26073]
### Current behavior

I have just upgraded from Cypress 9.7.0 to 12.7.0 and noticing that `before` hook runs before every test instead of once per `describe` block

I have also noticed (and may be this is related) that the Cypress runner keeps refreshing after each test 

Running the code with `CYPRESS_BASE_URL=<my url> cypress open`  and running one spec at a time 

Some observations:
1. when I set `testIsolation: false`, `before` hook now runs as expected
2. This seems to be an issue when I have both `before` and `beforeEach` hooks. 

### Desired behavior

Run `before` hook once per `describe` block

### Test code to reproduce

```
describe('Sell flow', function () {
    before(() => {
        cy.loginAndSetup(ENV_NAME);
        
        //Enabling feature can only be done after login, but only needs to be done once 
        cy.enableFeature(featureX);
    });

    beforeEach(function () {
        ... few intercepts setup 
    });

    afterEach(function () {
       if (this.currentTest.state === 'failed' && this.currentTest.currentRetry() == 1) {
           Cypress.runner.stop();
        }
    });

    it('Test 1, function () {
        .... some asserts here 
    });

    it('Test 2', function () {
        ... some asserts here 
    });
});
```

### Cypress Version

12.7.0

### Node version

v16.13.2

### Operating System

macOS Ventura 13.1

**Comment by mike-plummer** [comment]
@elenasch Sorry to hear you're having issues. I set up a minimal project with what you've described and the hooks appear to be executing as expected, so there must be something unique in your project. Can you provide a reproduction case that demonstrates the issue or provide more insight into what is being done in your custom commands that might be causing the test runner to reinitialize? Are there any plugins or customizations in use?

**_`spec.cy.js`_**
```js
describe('my suite', function () {
  before(() => {
      cy.log('`before`')
  });

  beforeEach(function () {
    cy.log('`beforeEach`')
  });

  afterEach(function () {
    cy.log('`afterEach`')
  });

  after(function () {
    cy.log('`after`')
  })

  it('Test 1', function () {});

  it('Test 2', function () {});

  it('Test 3', function () {});
});
```

**_Output_**

![Image](https://user-images.githubusercontent.com/11482842/224811253-63929fc2-f214-412a-a422-b1ab25feba85.png)

**Comment by Bhargavigundu** [comment]
The same scenario am also facing with before and beforeEach hook. With initial run it's working as expected like running before hook only once a spec file and beforeEach for every test with cypress runner and UI. But while executing from next time  it's running before hook as well for every test. And as mentioned in issue cypress UI as well getting refreshed for every test from 2nd run.

**Comment by mike-plummer** [comment]
@Bhargavigundu Unfortunately we have not received a reproduction case for this issue so there isn't much we can do. If you could provide one we'd be happy to try to assist with diagnosing the problem - this feature is working as expected for the vast majority of Cypress users so it is likely you have a configuration issue or misbehaving plugin in your project. Here are some tips for providing a [Short, Self Contained, Correct, Example](http://sscce.org/) and our own [Troubleshooting Cypress](https://on.cypress.io/troubleshooting) guide.

**Comment by pateltanay** [comment]
Has anyone found the solution to this issue? I am having the same issue, it is not happening when you just log a message or load fixture, but happening when using cy.visit() in in beforeall block. Any suggestions appreciated.

**Comment by lmiller1990** [comment]
Can you share a minimal reproduction @pateltanay? If so, you could open a new issue with it, and someone will take a look.

**Comment by pateltanay** [comment]
@lmiller1990 I was able to get past this with visit main domain before using cy.origin(). Thank you for taking the look.

**Comment by rawestmoreland** [comment]
I am also experience this. I have a before() that runs in my describe block with a custom command to log in with auth0. Yes, it's running before every test block instead of just once.

```
/* eslint-disable no-shadow */
/* eslint-disable sort-keys */

function loginViaAuth0Ui(username: string, password: string) {
  // App landing page redirects to Auth0.
  cy.visit('/');

  cy.get('a:contains(Login)').click({ force: true });

  // Login on Auth0.
  cy.origin(
    Cypress.env('auth0_domain'),
    { args: { username, password } },
    ({ username, password }) => {
      cy.get('input#username').type(username);
      cy.get('input#password').type(password, { log: false });
      cy.contains('button[value=default]', 'CONTINUE').click({force: true});
    }
  )

  // Ensure Auth0 has redirected us back to the RWA.
  cy.url().should('equal', `${Cypress.config().baseUrl}/deals`)
}

// @ts-ignore
Cypress.Commands.add('loginToAuth0', (username: string, password: string) => {
  const log = Cypress.log({
    // @ts-ignore
    autoEnd: false,
    displayName: 'AUTH0 LOGIN',
    message: [`🔐 Authenticating | ${username}`],
  });
  log.snapshot('before');

  loginViaAuth0Ui(username, password);

  log.snapshot('after');
  log.end();
});

export {};
```

```
describe('Empty destination row state', () => {
  before(() => {
    cy.loginToAuth0(Cypress.env('auth0_username_empty_premium'), Cypress.env('auth0_password_empty_premium'));
  });

  after(() => {
    cy.visit('/logout');
  });

  it('Has an empty watchlist state', () => {
    cy.visit('/watchlist');
    cy.url().should('equal', `${Cypress.config().baseUrl}/watchlist`);
    cy.contains('Your Watchlist is Empty');
  });

  it('Should navigate to explore', () => {
    cy.visit('/watchlist');
    cy.get('a:contains(Explore Deals)').click();
    cy.url().should('equal', `${Cypress.config().baseUrl}/deals`);
  });
});
```

**Comment by lmiller1990** [comment]
We will need a minimal, self contained example to reproduce this. I tried to put something together based on the above but there's too many things missing. It's too difficult and time consuming to piece together things based on a few snippets, considering a bunch of other requirements are needed.

I will close this until we have more to go on. I will reopen and look into this if we can have a minimal reproduction.

**Comment by yelliver** [comment]
I also got this error with version 12.17.3, this is a minimal example:

```
describe("my suite", function () {
  before(() => {
    cy.log("`before`");
  });

  it("Test 1", function () {
    cy.visit("https://www.example.com/");
  });

  it("Test 2", function () {
    cy.visit("https://www.wikipedia.com/");
  });
});
```

![image](https://github.com/cypress-io/cypress/assets/1596947/21892747-6911-4b3a-bbea-21a9e89d89c3)

**Comment by MikeMcC399** [comment]
@yelliver 

I can confirm your findings that `BEFORE ALL` is repeated. I suspect this is due to `cy.visit` changing domains between Test 1 and Test 2.

If `cy.visit()` is replaced by `cy.request()` in your example, then `BEFORE ALL` only shows up once in the logs:

![image](https://github.com/cypress-io/cypress/assets/66998419/22220d5a-70bd-42a1-ac45-e5eaa9c3c38d)

@lmiller1990 

If you paste the test from @yelliver into a default installation of Cypress 12.17.3 set up for e2e you have a minimal repro setup.

**Comment by MikeMcC399** [comment]
- Repro created on https://github.com/MikeMcC399/cypress-test-tiny/blob/repeated-before

- Instructions on https://github.com/MikeMcC399/cypress-test-tiny/blob/repeated-before/README.md

**Comment by nagash77** [comment]
Hi everyone,  I have reviewed this issue with the team and this is intended behavior.  If the `before( )` block was not executed on new navigation, all configuration done in it would be lost on navigation.

**Comment by MikeMcC399** [comment]
Hi Ben @nagash77 

Thanks very much for clarifying!

Maybe this should be added as a documentation note to [`cy.visit()`](https://docs.cypress.io/api/commands/visit)?

- [cy.visit() > Visiting cross-origin sites](https://docs.cypress.io/api/commands/visit#Visiting-cross-origin-sites)

That could extend the current statement "When visiting a cross-origin site, the `onBeforeLoad` and `onLoad` options are not supported.".
